contracts/linear/src/epoch_actions.rs

```diff
@@ -204,6 +204,137 @@ impl LiquidStakingContract {
+        self.assert_manager();
+
+        // make sure enough gas was given
+        let min_gas = GAS_DRAIN_UNSTAKE + GAS_EXT_UNSTAKE + GAS_CB_VALIDATOR_UNSTAKED;
+        require!(
+            env::prepaid_gas() >= min_gas,
+            format!("{}. require at least {:?}", ERR_NO_ENOUGH_GAS, min_gas)
+        );
+
+        let mut validator = self.validator_pool.get_validator(&validator_id)
+            .expect(ERR_VALIDATOR_NOT_EXIST);
+
+        // make sure the validator:
+        // 1. has weight set to 0
+        // 2. not in pending release
+        // 3. has not unstaked balance (because this part is from user's unstake request)
+        require!(
+            validator.weight == 0,
+            ERR_NON_ZERO_WEIGHT
+        );
+        require!(
+            !validator.pending_release(),
+            ERR_VALIDATOR_UNSTAKE_WHEN_LOCKED
+        );
+        require!(
+            validator.unstaked_amount == 0,
+            ERR_NON_ZERO_UNSTAKED_AMOUNT
+        );
+
+        let unstake_amount = validator.staked_amount;
+
+        Event::DrainUnstakeAttempt {
+            validator_id: &validator_id,
+            amount: &U128(unstake_amount)
+        }
+        .emit();
+
+        // perform actual unstake
+        validator
+            .unstake(&mut self.validator_pool, unstake_amount)
+            .then(ext_self_action_cb::validator_drain_unstaked_callback(
+                validator.account_id,
+                unstake_amount,
+                env::current_account_id(),
+                NO_DEPOSIT,
+                GAS_CB_VALIDATOR_UNSTAKED,
+            ));
+    }
+
+    /// Withdraw from a drained validator
+    pub fn drain_withdraw(&mut self, validator_id: AccountId) {
+        self.assert_manager();
+
+        // make sure enough gas was given
+        let min_gas = GAS_DRAIN_WITHDRAW + GAS_EXT_WITHDRAW + GAS_CB_VALIDATOR_WITHDRAW;
+        require!(
+            env::prepaid_gas() >= min_gas,
+            format!("{}. require at least {:?}", ERR_NO_ENOUGH_GAS, min_gas)
+        );
+
+        let mut validator = self
+            .validator_pool
+            .get_validator(&validator_id)
+            .expect(ERR_VALIDATOR_NOT_EXIST);
+
+        // make sure the validator:
+        // 1. has weight set to 0
+        // 2. has no staked balance
+        // 3. not pending release
+        require!(
+            validator.weight == 0,
+            ERR_NON_ZERO_WEIGHT
+        );
+        require!(
+            validator.staked_amount == 0,
+            ERR_NON_ZERO_STAKED_AMOUNT
+        );
+        require!(
+            !validator.pending_release(),
+            ERR_VALIDATOR_WITHDRAW_WHEN_LOCKED
+        );
+
+        let amount = validator.unstaked_amount;
+
+        Event::DrainWithdrawAttempt {
+            validator_id: &validator_id,
+            amount: &U128(amount)
+        }
+        .emit();
+
+        validator.withdraw(&mut self.validator_pool, amount).then(
```

Comment:
An alternative approach is to call `withdraw_all()` on the staking pool, but I think it's OK to specify the amount as long as there's no precision loss when calculating the amount.